libr/debug/debug.c

```diff
@@ -1145,6 +1105,45 @@ R_API int r_debug_continue_kill(RDebug *dbg, int sig) {
 	if (!dbg) {
 		return 0;
 	}
+
+	if (dbg->session) {
+		// If the debugger is not at the end of the session
+		// Go to the end or the next breakpoint
+		if (dbg->session->cnum != dbg->session->maxcnum) {
+			bool has_bp = false;
+			RRegItem *ripc = r_reg_get (dbg->reg, dbg->reg->name[R_REG_NAME_PC], R_REG_TYPE_GPR);
+			RVector *vreg = ht_up_find (dbg->session->registers, ripc->offset | (ripc->arena << 16), NULL);
+			RDebugChangeReg *reg;
+			r_vector_foreach_prev (vreg, reg) {
+				if (reg->cnum <= dbg->session->cnum) {
+					continue;
+				}
+				has_bp = r_bp_get_in (dbg->bp, reg->data, R_BP_PROT_EXEC) != NULL;
+				if (has_bp) {
+					eprintf ("hit breakpoint at: 0x%" PFMT64x " cnum: %d\n", reg->data, reg->cnum);
+					r_debug_goto_cnum (dbg, reg->cnum);
+					return dbg->tid;
+				}
+			}
+
+			r_debug_goto_cnum (dbg, dbg->session->maxcnum);
+			return dbg->tid;
+		}
+
+		if (dbg->trace_continue) {
+			// Continuously stepping to trace each instruction until user interrupt ^C
+			r_cons_break_push (NULL, NULL);
+			while (!r_cons_is_breaked ()) {
+				r_debug_step (dbg, 1);
```

Comment:
You should also check r_debug_step's return value, it returns the number of steps done on success.